### PR TITLE
use memory pool for storeDicts

### PR DESCRIPTION
**Pull request by metalogical**
This is important to reduce GC pressure in applications that spend a lot of time parsing. Let me know if you want more details on this, but we're already using this on our end, so would be great if we can merge it upstream.

@mna @breml 

EDIT: also let me know if you think we should write benchmarks to verify this

### CODE REVIEW

**Comment by breml** [comment]
@metalogical Thanks for your PR. The PR looks pretty straight forward to me and it makes sense, that if a long running process uses the parse a lot, there is some pressure on the GC. I would really appreciate, if you could provide some benchmarks, that show the positive effect of the `sync.Pool` in that case. Additionally it would be interesting to investigate, if there are other spots, where maps are used, that could benefit from `sync.Pool` as well.

File: builder/static_code.go [hunk 1]
```diff
@@ -722,6 +723,17 @@ type Cloner interface {
 	Clone() interface{}
 }
 
+var statePool = &sync.Pool{
+	New: func() interface{} { return make(storeDict, 2) },
```
Comment (breml) [inline builder/static_code.go]: Why is the `storeDict` initialized with a size of `2`? is there a particular reason for this?

Comment (hrysoula) [reply builder/static_code.go]: Good catch, this was an optimization attempt that is not performant.  Fixed in 2ee1908.

**Comment by hrysoula** [comment]
> Please run make clean && make all. The file bootstrap/cmd/bootstrap-pigeon/bootstrap_pigeon.go is not up to date in the last commit.

Fixed in e25b068.

**Comment by hrysoula** [comment]
Internally we have run benchmarks that verify the improvement. I unfortunately cannot provide the parser as it is a private repository.

Before:
```
goos: linux
goarch: amd64
pkg: github.com/kiteco/kiteco/kite-go/lang/python/pythonparser/calls/internal/pigeon
BenchmarkValidStrings-4     	     500	   3367479 ns/op
BenchmarkInvalidStrings-4   	   10000	    180598 ns/op
PASS
```

After:
```
goos: linux
goarch: amd64
pkg: github.com/kiteco/kiteco/kite-go/lang/python/pythonparser/calls/internal/pigeon
BenchmarkValidStrings-4     	     500	   2808122 ns/op
BenchmarkInvalidStrings-4   	   10000	    131682 ns/op
PASS
```

Running the existing [`benchparse_test.go`](https://github.com/mna/pigeon/blob/master/benchparse_test.go) tests, the improvement is less noticeable but still present:

Before:
```
goos: darwin
goarch: amd64
pkg: github.com/mna/pigeon
BenchmarkParseUnicodeClass-8   	   10000	    216224 ns/op
BenchmarkParseKeyword-8        	   10000	    135896 ns/op
PASS
```

After:
```
goos: darwin
goarch: amd64
pkg: github.com/mna/pigeon
BenchmarkParseUnicodeClass-8   	   10000	    212132 ns/op
BenchmarkParseKeyword-8        	   10000	    135354 ns/op
PASS
```

**Comment by metalogical** [comment]
@breml are the existing benchmarks that Hrysoula posted above enough, or would you like something more to get this merged?

cc @mna

**Comment by breml** [comment]
@hrysoula Thanks for the benchmarks. These look good to me. So I will merge this.

**Pull request merged**